tests/components/aosmith/test_init.py

```diff
@@ -0,0 +1,88 @@
+"""Tests for the initialization of the A. O. Smith integration."""
+
+from datetime import timedelta
+from unittest.mock import MagicMock, patch
+
+from py_aosmith import AOSmithInvalidCredentialsException, AOSmithUnknownException
+import pytest
+
+from homeassistant.components.aosmith.const import (
+    DOMAIN,
+    FAST_INTERVAL,
+    REGULAR_INTERVAL,
+)
+from homeassistant.config_entries import ConfigEntryState
+from homeassistant.core import HomeAssistant
+import homeassistant.util.dt as dt_util
+
+from tests.common import MockConfigEntry, async_fire_time_changed
+
+
+async def test_config_entry_setup(init_integration: MockConfigEntry) -> None:
+    """Test setup of the config entry."""
+    mock_config_entry = init_integration
+
+    assert mock_config_entry.state is ConfigEntryState.LOADED
+
+
+async def test_config_entry_not_ready(
+    hass: HomeAssistant, mock_config_entry: MockConfigEntry
+) -> None:
+    """Test the config entry not ready."""
+    mock_config_entry.add_to_hass(hass)
+
+    with patch(
+        "homeassistant.components.aosmith.config_flow.AOSmithAPIClient.get_devices",
+        side_effect=AOSmithUnknownException("Unknown error"),
+    ):
+        await hass.config_entries.async_setup(mock_config_entry.entry_id)
+        await hass.async_block_till_done()
+
+    assert mock_config_entry.state is ConfigEntryState.SETUP_RETRY
+
+
+async def test_update_auth_failure(
+    hass: HomeAssistant, mock_client: MagicMock, init_integration: MockConfigEntry
+) -> None:
+    """Test auth failure during data update."""
+    entries = hass.config_entries.async_entries(DOMAIN)
+    assert len(entries) == 1
+    assert entries[0].state is ConfigEntryState.LOADED
+
+    mock_client.get_devices.side_effect = AOSmithInvalidCredentialsException(
+        "Authentication error"
+    )
+    async_fire_time_changed(hass, dt_util.utcnow() + REGULAR_INTERVAL)
```

Comment:
Please use the freezer for changing time

Reply:
I actually tried to use the freezer at first, but one of the tests fails when I try to use it. Actually, even if I just add the freezer as a parameter to this test, the test fails, even if I don't change the implementation at all. I have no idea what's causing this.

Reply:
Sounds interesting, iirc we should use the freezer for any time related test. If you want to discuss it, feel free to reach out on discord

Reply:
I figured out the problem. It was caused by the order of the parameters of the test. I was adding the freezer parameter after the `init_integration` parameter, and the test was failing. When I moved it to before `init_integration`, it fixed it. I think it's because the freezer fixture advances time as part of its initialization, which was causing the coordinator to update method to be called when it shouldn't. Anyway, it's fixed now.